doc/design/upgrade.md

```diff
@@ -125,15 +135,114 @@ Here is the upgrade workflow
+
+During appliance upgrade, vic-machine should have clear information to mention the container functional limitation after upgrade, and the solution for it.
+
+### KeyValue Store Migration
+Different to guestinfo object, VIC will not change configuration from guestinfo to anywhere else anytime soon. But from start, keyvalue store persistent position is in argument. Right now, vic has a few datastore files for keyvalue store, but those information is not configurable, which means hardcoded in portlayer.
+
+In the future, if there is any change on assumption no matter for what reason, performance, easy to management, data migration will be hard to do.
+
+So the idea is that, data migration framework does not assume where to load keyvalue store. If there is keyvalue store changes, plugin is the owner to read old version, update to new version and then write back to datastore files, or to anywhere defined in new version, suppose the configuration can be retrieved from guestinfo. So the input of keyvalue store plugin will be same to other appliance configuration update plugin.
+
+The problem of this solution is that each single plugin will persist its own change, not like the guestinfo update, which is migrated in memory and persisted by migration framework. So the rollback for guestinfo is easy, but not possible for keyvalue store plugin, unless we have data roll back plugin mechanism.
+
+The workaround for this idea is to restrict keyvalue store migration functions, e.g. only new key is supported. In the plugin, do not remove old keyvalues, so in the above failure case, old binary can still work with old keyvalues.
+
+We have lots of other options to better support keyvalue store upgrade, but at this time, I don't want to predicate our changes in the future, that's easy to be over-engineering. While that happens, we could think about to support plugin roll back, or upgrade from portlayer, etc.
+
+### Appliance log files
+
+Appliance log files configuration will not be touched at this time. So upgraded VCH will use old log files if there is any such kind of change. We can revisit this while there is customer requiement on this.
+
+### vSphere Object Management Logic
+
+There will have new vSphere API come up, so the logic to manage vsphere objects will be enhanced, for example, as Caglar said the vmdk can be managed directly in vSphere 6.5, instead of through vm operations in vSphere 6.0. We'll switch to these new interfaces for image and volume management sometime later.
+
+vic-machine is not supposed to migrate old image data or volume data to new vmdk files, so portlayer will need to be backward compatible. (Suppose vSphere will be backward compatibile, there should be nothing to do in vic, but need to do some research on it)
+
+### How to Support Upgrade Between Builds
```

Comment:
> 1. Why do we call a migration patch as a plugin? Will it be stored as a separate file on disk?

The plugin name is corresponding to upgrade framework. Each plugin is one migration function called by framework based on their version ID.

> 2. ...

We're doing that based on new data migration version. Which is not related to release version. The PR #3478 has one prototype for this idea. You might check that to see if there is any inconsistency with your suggestion.

> 3. 

We did backup guestinfo in current upgrade code, and in fatal upgrade, everything could be reverted.
But the only exception is keyvalue store, which is described in previous section.